### PR TITLE
test/refactor: update isExternalLink

### LINKED ISSUE #3833: Regression in CI benchmark
Usually cold processing can complete within 25 seconds, lately I noticed it has increased to more than a minute.

I haven't find the cause yet, I did notice the regression in https://github.com/hexojs/hexo/pull/3808 https://github.com/hexojs/hexo/pull/3812 https://github.com/hexojs/hexo/pull/3813

Commit Time (GMT) | Commit | Cold Processing Time<br/>(Node 8)
--- | --- | ---
#3813
27 Oct 14:48 | c8fee535b310528fdc01a7761e5df5edf91f9c8c | 24s
29 Oct 12:26 | 330d88bf3a59a949a94677a81d2dc8619ba78ab5 | 38s
#3808
28 Oct 06:39 | 28949da9412f9ca8e08c91b85a73c1e7cbfc24fc | 23s
02 Nov 03:29 | 48d2c527370800c25c6093be8fd857e70c9c0045 | 40s
#3812 
27 Oct 13:32 | 322251361950463e16c1fbeac6e9bc5c995dbc6e | 23s
28 Oct  05:09 | ab23935be17f7b2bc6bd79488c1a065f164e5398 | 35s

Reverting #3808 and #3813  in #3831, but no difference. Haven't try #3812 yet.

Commit Time (GMT) | Commit | PR | Cold Processing Time<br/>(Node 8)
--- | --- | --- | ---
Latest master
05 Nov 01:32 | 77f3e21c34cec3de72409989c2babea582594f03 | #3829 | 42s
Revert attempt
05 Nov 02:21 | 8a475a845f70151521fae0a3027415642e95c499 | #3831 | 40s

---

I also tried dropping recent commits until 3448a6dccdcc4fcb264514e3ffe8f11060a00d3e, see [revert-lodash](https://github.com/curbengh/hexo/tree/revert-lodash) tree. The processing time did come back to normal 26s ([build log](https://travis-ci.com/curbengh/hexo/jobs/252867495)).

---

Looks like whatwg API is the cause #3812 
I revert #3812 in https://github.com/hexojs/hexo/pull/3834, processing time is 27s ([build log](https://travis-ci.com/hexojs/hexo/jobs/252876628)).

In Node 13, it dropped from [81s](https://travis-ci.com/hexojs/hexo/jobs/252852712) to [27s](https://travis-ci.com/hexojs/hexo/jobs/252876631).

### PR DESCRIPTION
**Pull request by SukkaW**
## What does it do?

Try to fix #3833 

https://github.com/hexojs/hexo/issues/3833#issuecomment-549689276

## Pull request tasks

- [ ] Add test cases for the changes.
- [x] Passed the CI test.

### CODE REVIEW

**Comment by SukkaW** [comment]
After https://github.com/hexojs/hexo/pull/3839/commits/60acb95c4c624fd904531103520cd2a02eec494f , I managed to lower the generation duration from `86s` to `76s` on Node.js 12 & 13.

**Comment by SukkaW** [comment]
From https://github.com/MaxCDN/bootstrapcdn/pull/1391 , the matainer of bootstrapcdn said the `url.parse` is faster. Maybe we should have a benchmark on which one is faster.

**Comment by SukkaW** [comment]
According to the [benchmark](https://runkit.com/sukkaw/5dc278520410ce00133cb012), the current "safe whatwg url" or `urlObj` is 2x slower than pure `new URL`, 6x slower than `url.parse()`.

Since whatwg url is required in `isExternalLink` to ignore `mailto:` and so on, I might pass `sitehost` as base in `new URL` to avoid `urlObj`.

**Comment by SukkaW** [comment]
The generation duration is dropped to:

- Node.js 10: `27s`
- Node.js 12: `26s`
- Node.js 13: `24s`

**Comment by SukkaW** [comment]
@curbengh 

It is now ready for review.

After this PR is approved, we should close this PR and update `isExternalLink` in `hexo-util`.

File: lib/plugins/filter/after_render/external_link.js [hunk 1]
```diff
@@ -51,10 +43,13 @@ function externalLinkFilter(data) {
   if (config.external_link === false || config.external_link.enable === false
     || config.external_link.field !== 'site') return;
 
+  config.external_link.exclude = Array.isArray(config.external_link.exclude) ? config.external_link.exclude
+    : [config.external_link.exclude];
+
   data = data.replace(/<a.*?(href=['"](.*?)['"]).*?>/gi, (str, hrefStr, href) => {
-    if (/target=/gi.test(str) || !isExternal(href, config)) return str;
+    if (str.includes('target=') || !isExternal(href, config)) return str;
```
Comment (curbengh) [inline lib/plugins/filter/after_render/external_link.js]: `String.includes()` is case-sensitive, so this is not equivalent to the `i` regexp flag, need `str.toLowerCase().includes()`. Hopefully `String.toLowerCase()` is not expensive.

Comment (SukkaW) [reply lib/plugins/filter/after_render/external_link.js]: According the [benchmark](https://runkit.com/sukkaw/5dc6398fcb31010019d500e6), `test()` is faster. I will revert `includes()` then,

**Comment by curbengh** [comment]
Superseded by https://github.com/hexojs/hexo/pull/3847.

**Pull request closed without merging**